docs/guides/dropout.rst

```diff
@@ -0,0 +1,147 @@
+that is splittable.
+
+**Note:** Recall that JAX has an explicit way of giving you PRNG keys:
+you can fork the main PRNG state (such as ``key = jax.random.PRNGKey(seed=0)``)
+into multiple new PRNG keys with ``key, subkey = jax.random.split(key)``. You
+can refresh your memory in
+`🔪 JAX - The Sharp Bits 🔪 Randomness and PRNG keys <https:/t/jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers>`__.
+
+You split the PRNG into keys, including one for dropout, as follows::
+
+  seed = 0
+  root_key = jax.random.PRNGKey(seed=seed)
+  main_key, params_key, dropout_key = jax.random.split(key=root_key, num=3)
+
+Each of the Flax PRNG streams has a name, such as the ``'params'`` stream for
+initializing parameters, and ``'dropout'`` for dropout.
+
+Defining a model with ``Dropout``
+*********************************
+
+To create a model, subclass :meth:`flax.linen.Module`, and then use
+:meth:`flax.linen.Dropout` to add a dropout layer. Recall that
+:meth:`flax.linen.Module` is the
+`base class for all neural network modules <https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module>`__
+, and all layers and models are subclassed from it.
+
+In :meth:`flax.linen.Dropout`, the ``deterministic`` argument is ``None`` by
+default.
+
+* If it's set to ``False``, the inputs are scaled by ``1 / (1 - dropout_rate)`` and masked.
+* When it's ``True``, no mask is applied (the dropout is turned off).
+
+Flax provides an implicit way of handling PRNG key streams via Flax
+:meth:`flax.linen.Module`'s :meth:`flax.linen.Module.make_rng` helper function.
+It allows the code in Flax ``Module``s (or their sub-``Module``s) to
+"pull PRNG keys". ``make_rng`` guarantees to provide a unique key each time you
+call it.
+
+**Note:** Internally, :meth:`flax.linen.Dropout` makes use of :meth:`flax.linen.Module.make_rng`
+to create a key for dropout. You can check out the
+`source code <https://github.com/google/flax/blob/5714e57a0dc8146eb58a7a06ed768ed3a17672f9/flax/linen/stochastic.py#L72>`__.
+
+* Every time :meth:`Module.make_rng` is called (which is done implicitly in
+  :meth:`flax.linen.Dropout`), you get a new PRNG key split from the main/root pseudorandom
+  number generator (PRNG) key.
+* :meth:`Module.make_rng` guarantees full reproducibility.
+
+.. codediff::
+  :title_left: No Dropout
+  :title_right: With Dropout
+  :sync:
+
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+
+
+
+      return x
+
+  ---
+  class MyModel(nn.Module):
+    num_neurons: int
+    training: bool
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(self.num_neurons)(x)
+      # Set the dropout layer with a rate of 50%. #!
+      # When the `deterministic` flag is `True`, dropout is turned off. #!
+      x = nn.Dropout(rate=0.5, deterministic=not self.training)(x) #!
```

Comment:
Ideally we keep passing training state via a call argument as the user can control it via `Module.apply` / `TrainState.apply_fn`, else you'd be forced to either use mutable state (which has pitfalls) or recreating entire `Module` / `TrainState`.
```suggestion

    @nn.compact
    def __call__(self, x, train: bool):
      x = nn.Dense(self.num_neurons)(x)
      # Set the dropout layer with a rate of 50%. #!
      # When the `deterministic` flag is `True`, dropout is turned off. #!
      x = nn.Dropout(rate=0.5, deterministic=not train)(x) #!
```

Reply:
@cgarciae Can we remind users that they can use both `train` and `training`? This may start getting confusing especially for less experienced users. cc @levskaya @marcvanzee (API design)

Reply:
> Ideally we keep passing training state via a call argument as the user can control it via Module.apply / TrainState.apply_fn, else you'd be forced to either use mutable state (which has pitfalls) or recreating entire Module / TrainState.

Done. Thanks @cgarciae